src/software/util/design_patterns/constructor_arg_generic_factory.h

```diff
@@ -0,0 +1,187 @@
+template <class IndexType, class TypeToCreate>
+using ConstructorArgGenericRegistry =
+std::unordered_map<IndexType, std::function<std::unique_ptr<TypeToCreate>(std::any)>>;
+/**
+ * The GenericFactory is an Abstract class that provides an interface for Generic type
+ * Factories to follow. This makes it easy to maintain a list of factories and get the
+ * corresponding generic types through the generic interface.
+ */
+
+// had to roll our own function_traits class to get function arg types as a tuple
+template <typename Function> class fn_traits;
+
+template <typename ResultType, typename... ArgTypes>
+class fn_traits<ResultType(ArgTypes...)>
+{
+   public:
+    using args_tuple_type = std::tuple<ArgTypes...>;
+    using result_type = ResultType;
+};
+
+
+// modified from std::make_from_tuple in libstdc++ with names de-uglified
+template <typename CreatedType, typename Tuple, size_t... Idx>
+std::unique_ptr<CreatedType> make_unique_from_tuple_impl(Tuple&& tup, std::index_sequence<Idx...>)
+{ return std::unique_ptr<CreatedType>(new CreatedType(std::get<Idx>(std::forward<Tuple>(tup))...)); }
+
+template <typename CreatedType, typename Tuple>
+std::unique_ptr<CreatedType> make_unique_from_tuple(Tuple&& tup)
+{
+    return make_unique_from_tuple_impl<CreatedType>(
+        std::forward<Tuple>(tup),
+        std::make_index_sequence<std::tuple_size_v<std::decay_t<Tuple>>>{});
+}
+
+template <class IndexType, class TypeToCreate>
+class ConstructorArgGenericFactory
+{
+   public:
+    /**
+     * Returns a unique pointer to a newly constructed type of the given type/name
+     *
+     * @param generic_name The name of the type to construct. This value must be in the
+     * Generic registry
+     * @throws std::invalid_argument if the given generic_name is not found in the Generic
+     * registry
+     *
+     * @return a unique pointer to a newly constructed type of the given type/name
+     */
+    template <typename U>
+    static std::unique_ptr<TypeToCreate> create(const std::string& generic_name, U constructor_arg);
+
+    static std::unique_ptr<TypeToCreate> create(const std::string& generic_name, std::any constructor_arg);
+
+    template <typename ... ArgTypes>
+    static std::unique_ptr<TypeToCreate> create(const std::string& generic_name, ArgTypes... args);
+
+    /**
+     * Returns a const reference to the generic type registry. The registry is a map of
+     * generic names to a "create" function that will create and return a unique_ptr to a
+     * new concrete instance of the generic type.
+     *
+     * @return a const reference to the generic registry
+     */
+    static const ConstructorArgGenericRegistry<IndexType, TypeToCreate>& getRegistry();
+
+    /**
+     * Returns the list of names for all creator functions registered in the factory
+     *
+     * @return a list of names for all creator functions registered in the factory
+     */
+    static std::vector<std::string> getRegisteredNames();
```

Comment:
Should this return a `std::vector<IndexType>` instead? In which case the function name should probably also change.